pkg/storage/credentialset.go

```diff
@@ -77,13 +86,63 @@ func (s CredentialSet) DefaultDocumentFilter() map[string]interface{} {
 	return map[string]interface{}{"namespace": s.Namespace, "name": s.Name}
 }
 
-func (s CredentialSet) Validate() error {
-	if CredentialSetSchemaVersion != s.SchemaVersion {
-		if s.SchemaVersion == "" {
-			s.SchemaVersion = "(none)"
+func (s *CredentialSet) Validate(ctx context.Context, strategy schema.CheckStrategy) error {
+	ctx, span := tracing.StartSpan(ctx,
+		attribute.String("credentialset", s.String()),
+		attribute.String("schemaVersion", string(s.SchemaVersion)),
+		attribute.String("defaultSchemaVersion", string(DefaultCredentialSetSchemaVersion)))
+	defer span.EndSpan()
+
+	// Before we can validate, get our resource in a consistent state
+	// 1. Check if we know what to do with this version of the resource
+	if warnOnly, err := schema.ValidateSchemaVersion(strategy, SupportedCredentialSetSchemaVersions, string(s.SchemaVersion), DefaultCredentialSetSemverSchemaVersion); err != nil {
+		if warnOnly {
+			span.Warn(err.Error())
+		} else {
+			return span.Error(err)
+		}
+	}
+
+	// 2. Check if they passed in the right resource type
+	if s.SchemaType != "" && !strings.EqualFold(s.SchemaType, SchemaTypeCredentialSet) {
+		return span.Errorf("invalid schemaType %s, expected %s", s.SchemaType, SchemaTypeCredentialSet)
+	}
+
+	// 3. Migrate their resource to the most recently supported version if necessary
+	if err := s.Migrate(ctx); err != nil {
+		return err
+	}
+
+	// Default the schemaType before importing into the database if it's not set already
+	// SchemaType isn't really used by our code, it's a type hint for editors, but this will ensure we are consistent in our persisted documents
+	if s.SchemaType == "" {
+		s.SchemaType = SchemaTypeCredentialSet
+	}
+
+	// OK! Now we can do resource specific validations
+	return nil
+}
+
+// Migrate checks if the resources is in an older schema version, and if so applies successive migrations until it is up-to-date.
+func (s *CredentialSet) Migrate(ctx context.Context) error {
+	//lint:ignore SA4006 ignore unused ctx for now
+	ctx, span := tracing.StartSpan(ctx)
+	defer span.EndSpan()
+
+	for s.SchemaVersion != DefaultCredentialSetSchemaVersion {
+		switch s.SchemaVersion {
```

Comment:
nit: This migration (and the other resource migration code) seems like it could get unwieldy quickly and become a bit of a maintenance headache having all of this switch logic here. Is there a way to simplify schema migration for resources using like the strategy pattern or something along those lines?